File: code/nodejs/05-context-assembly/context_assembler.ts
/**
 * Context Assembler — dynamic context assembly from multiple sources.
 *
 * Assembles the LLM context string from:
 *   1. RAG retrieved documents
 *   2. Tool execution results
 *   3. User profile / preferences
 *   4. Conversation summary
 *   5. Template variables
 *
 * Integrates with ContextBudget for enforcement and a lightweight
 * structure helper for attention-aware formatting.
 *
 * See: docs/04-context-engineering/01-the-context-window-as-a-resource.md
 */

import { ContextBudget, countTokens, type ContentInput } from "./context_budget.js";
import { get_encoding } from "tiktoken";

// ---------------------------------------------------------------------------
// Simple template renderer (${var} syntax)
// ---------------------------------------------------------------------------

function renderTemplate(template: string, vars: Record<string, string>): string {
  return template.replace(/\$\{(\w+)\}|\$(\w+)/g, (_, braced, bare) => {
    const key = braced ?? bare;
    return key in vars ? vars[key] : `$${key}`;
  });
}

// ---------------------------------------------------------------------------
// Lightweight structure helper (mirrors ContextOptimizer.structureForRetrieval)
// ---------------------------------------------------------------------------

function structureDocuments(documents: Array<{ text: string; metadata?: Record<string, unknown> }>): string {
  const parts: string[] = ["## Context Overview\n"];
  for (let i = 0; i < documents.length; i++) {
    const source = (documents[i].metadata?.["source"] as string) ?? `Document ${i + 1}`;
    parts.push(`- Section ${i + 1}: ${source}`);
  }
  parts.push("\n---\n");
  for (let i = 0; i < documents.length; i++) {
    const source = (documents[i].metadata?.["source"] as string) ?? `Document ${i + 1}`;
    parts.push(`## [${i + 1}] ${source}\n`);
    parts.push(documents[i].text);
    parts.push(`\n[End Section ${i + 1}]\n`);
    parts.push("---\n");
  }
  return parts.join("\n");
}

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

export interface ContextConfig {
  /** Python-style $variable or ${variable} template. */
  template: string;
  /** Variable substitutions for the template. */
  templateVars?: Record<string, string>;
  /** Which sources to include. Defaults to all four. */
  includeSources?: Array<"rag" | "tools" | "profile" | "summary">;
  /** Per-source token caps. */
  maxTokensPerSource?: Partial<Record<"rag" | "tools" | "profile" | "summary", number>>;
  /** Output format hint. Only "markdown" triggers structural optimisation. */
  format?: "markdown" | "plain" | "json";
  /** Preferred source order. */
  priorityOrder?: Array<"rag" | "tools" | "profile" | "summary">;
}

// ---------------------------------------------------------------------------
// Assembly result
// ---------------------------------------------------------------------------

export interface AssemblyResult {
  /** Final assembled context string. */
  context: string;
  /** Token count per source section. */
  tokenBreakdown: Record<string, number>;
  /** Sources that made it into the context. */
  sourcesIncluded: string[];
  /** Sources dropped due to budget exhaustion. */
  sourcesExcluded: string[];
  /** Sum of all included section tokens. */
  readonly totalTokens: number;
}

function makeAssemblyResult(
  context: string,
  tokenBreakdown: Record<string, number>,
  sourcesIncluded: string[],
  sourcesExcluded: string[]
): AssemblyResult {
  return {
    context,
    tokenBreakdown,
    sourcesIncluded,
    sourcesExcluded,
    get totalTokens() {
      return Object.values(this.tokenBreakdown).reduce((s, v) => s + v, 0);
    },
  };
}

// ---------------------------------------------------------------------------
// ContextAssembler
// ---------------------------------------------------------------------------

const SECTION_HEADERS: Record<string, string> = {
  rag:     "## Retrieved Documents\n",
  tools:   "## Tool Results\n",
  profile: "## User Profile\n",
  summary: "## Conversation Summary\n",
};

/**
 * Assemble context for LLM calls from multiple sources.
 *
 * @example
 * ```ts
 * const budget    = new ContextBudget(128_000);
 * const assembler = new ContextAssembler(budget);
 *
 * const result = await assembler.assemble({
 *   template: "You are a ${role} support agent.",
 *   variables: { role: "customer" },
 *   retrievedDocs: ragDocs,
 *   userProfile: { name: "Alice", plan: "Pro" },
 *   query: "How do I cancel?",
 * });
 * ```
 */
export class ContextAssembler {
  private readonly budget: ContextBudget;
  private readonly model: string;

  constructor(budget: ContextBudget, model: string = "gpt-4o") {
    this.budget = budget;
    this.model  = model;
  }

  /**
   * Assemble the dynamic context for an LLM call.
   *
   * Sources are assembled in priority order (rag → tools → profile → summary).
   * Each section is added while budget remains; oversized sections are truncated;
   * sections with no remaining budget are excluded.
   */
  assemble(options: {
    template: string;
    variables?: Record<string, string>;
    retrievedDocs?: Array<{ text: string; metadata?: Record<string, unknown> }>;
    toolResults?: Array<{ tool: string; result: unknown }>;
    userProfile?: Record<string, string>;
    conversationSummary?: string;
    query?: string;
    optimize?: boolean;
  }): AssemblyResult {
    const {
      template,
      variables = {},
      retrievedDocs,
      toolResults,
      userProfile,
      conversationSummary,
      query = "",
      optimize = true,
    } = options;

    // Render template
    const renderedTemplate = renderTemplate(template, variables);

    // Build source sections
    const sections: Record<string, string> = {};

    if (retrievedDocs && retrievedDocs.length > 0) {
      sections["rag"] =
        optimize && query
          ? structureDocuments(retrievedDocs)
          : retrievedDocs.map((d) => d.text).join("\n\n");
    }

    if (toolResults && toolResults.length > 0) {
      sections["tools"] = toolResults
        .map(({ tool, result }) => {
          const resultStr =
            typeof result === "object"
              ? JSON.stringify(result, null, 2)
              : String(result);
          return `**${tool}**:\n${resultStr}`;
        })
        .join("\n\n");
    }

    if (userProfile && Object.keys(userProfile).length > 0) {
      sections["profile"] = Object.entries(userProfile)
        .map(([k, v]) => `- **${k}**: ${v}`)
        .join("\n");
    }

    if (conversationSummary) {
      sections["summary"] = conversationSummary;
    }

    // Assemble within budget
    const dcBudget = this.budget.getTokenBudget("dynamic_context");
    const contextParts: string[] = [renderedTemplate];
    const tokenBreakdown: Record<string, number> = {
      template: countTokens(renderedTemplate, this.model),
    };
    const sourcesIncluded: string[] = [];
    const sourcesExcluded: string[] = [];

    for (const source of ["rag", "tools", "profile", "summary"] as const) {
      if (!(source in sections)) continue;

      const content    = sections[source];
      const header     = SECTION_HEADERS[source] ?? `## ${source}\n`;
      const sectionText  = header + content;
      const sectionTok   = countTokens(sectionText, this.model);
      const usedSoFar    = Object.values(tokenBreakdown).reduce((s, v) => s + v, 0);

      if (usedSoFar + sectionTok <= dcBudget) {
        contextParts.push(sectionText);
        tokenBreakdown[source] = sectionTok;
        sourcesIncluded.push(source);
      } else {
        const available = dcBudget - usedSoFar - countTokens(header, this.model);
        if (available > 100) {
          const truncated  = this.truncateToTokens(content, available);
          const truncText  = header + truncated;
          const truncTok   = countTokens(truncText, this.model);
          contextParts.push(truncText);
          tokenBreakdown[source] = truncTok;
          sourcesIncluded.push(source);
        } else {
          sourcesExcluded.push(source);
        }
      }
    }

    return makeAssemblyResult(
      contextParts.join("\n\n"),
      tokenBreakdown,
      sourcesIncluded,
      sourcesExcluded
    );
  }

  /**
   * Assemble context from a declarative {@link ContextConfig}.
   */
  assembleFromConfig(
    config: ContextConfig,
    data: {
      retrievedDocs?: Array<{ text: string; metadata?: Record<string, unknown> }>;
      toolResults?: Array<{ tool: string; result: unknown }>;
      userProfile?: Record<string, string>;
      conversationSummary?: string;
      query?: string;
    } = {}
  ): AssemblyResult {
    const included = new Set(config.includeSources ?? ["rag", "tools", "profile", "summary"]);

    let docs    = included.has("rag")     ? data.retrievedDocs      : undefined;
    const tools   = included.has("tools")   ? data.toolResults         : undefined;
    const profile = included.has("profile") ? data.userProfile         : undefined;
    const summary = included.has("summary") ? data.conversationSummary : undefined;

    // Apply per-source token caps
    const ragCap = config.maxTokensPerSource?.["rag"];
    if (ragCap != null && docs) {
      docs = this.clipDocsToBudget(docs, ragCap);
    }

    return this.assemble({
      template:             config.template,
      variables:            config.templateVars ?? {},
      retrievedDocs:        docs,
      toolResults:          tools,
      userProfile:          profile,
      conversationSummary:  summary,
      query:                data.query ?? "",
      optimize:             (config.format ?? "markdown") === "markdown",
    });
  }

  // ----------------------------------------------------------------
  // Helpers
  // ----------------------------------------------------------------

  private truncateToTokens(text: string, maxTokens: number): string {
    let enc;
    try {
      enc = get_encoding("cl100k_base");
    } catch {
      return text.slice(0, maxTokens * 4); // rough fallback
    }
    const tokens = enc.encode(text);
    if (tokens.length <= maxTokens) {
      enc.free();
      return text;
    }
    const result = new TextDecoder().decode(enc.decode(tokens.slice(0, maxTokens)));
    enc.free();
    return result;
  }

  private clipDocsToBudget(
    docs: Array<{ text: string; metadata?: Record<string, unknown> }>,
    maxTokens: number
  ): Array<{ text: string; metadata?: Record<string, unknown> }> {
    const kept: typeof docs = [];
    let used = 0;
    for (const doc of docs) {
      const t = countTokens(doc.text, this.model);
      if (used + t <= maxTokens) {
        kept.push(doc);
        used += t;
      } else {
        break;
      }
    }
    return kept;
  }
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

function runDemo(): void {
  console.log("=== Context Assembler (TypeScript) Demo ===\n");

  const budget    = new ContextBudget(16_000);
  const assembler = new ContextAssembler(budget);

  const ragDocs = [
    {
      text: "## Cancellation Policy\nYou may cancel at any time. Takes effect at end of billing period.",
      metadata: { source: "help-center/cancellation.md", score: 0.92 },
    },
    {
      text: "## Billing\nInvoices available at Settings → Billing → Invoice History.",
      metadata: { source: "help-center/billing.md", score: 0.89 },
    },
  ];

  const userProfile = {
    name:         "Alice Johnson",
    plan:         "Pro (annual)",
    member_since: "2022-03-15",
  };

  const result = assembler.assemble({
    template:            "You are a ${role} support agent. Answer using only the documents.\n",
    variables:           { role: "customer" },
    retrievedDocs:       ragDocs,
    userProfile,
    conversationSummary: "Alice asked about changing her password (resolved).",
    query:               "How do I cancel my subscription?",
  });

  console.log("Sources included :", result.sourcesIncluded);
  console.log("Sources excluded :", result.sourcesExcluded);
  console.log("Total tokens     :", result.totalTokens.toLocaleString());
  console.log("\nToken breakdown:");
  for (const [src, tok] of Object.entries(result.tokenBreakdown)) {
    console.log(`  ${src.padEnd(10)}: ${tok.toLocaleString()} tokens`);
  }
  console.log("\n--- Context (first 500 chars) ---");
  console.log(result.context.slice(0, 500));
}

// Run demo when executed directly
if (process.argv[1] && process.argv[1].endsWith("context_assembler.ts")) {
  runDemo();
}

File: code/python/05-context-assembly/context_assembler.py
"""Context Assembler — dynamic context assembly from multiple sources.

Assembles the LLM context string from:
  1. RAG retrieved documents
  2. Tool execution results
  3. User profile / preferences
  4. Conversation summary
  5. Template variables

Integrates with :class:`~context_budget.ContextBudget` for budget enforcement
and :class:`~context_optimizer.ContextOptimizer` for structure optimisation.

See: docs/04-context-engineering/01-the-context-window-as-a-resource.md
"""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass, field
from string import Template
from typing import Optional

import tiktoken

from context_budget import ContextBudget, EnforceResult, count_tokens
from context_optimizer import ContextOptimizer

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------


@dataclass
class ContextConfig:
    """Declarative configuration for context assembly.

    Example::

        config = ContextConfig(
            template="You are a $role. Use the documents to answer.",
            template_vars={"role": "support_agent"},
            include_sources=["rag", "profile"],
            max_tokens_per_source={"rag": 50_000},
            format="markdown",
        )
    """

    template: str
    template_vars: dict[str, str] = field(default_factory=dict)
    include_sources: list[str] = field(
        default_factory=lambda: ["rag", "tools", "profile", "summary"]
    )
    max_tokens_per_source: dict[str, int] = field(default_factory=dict)
    format: str = "markdown"   # "markdown" | "plain" | "json"
    priority_order: list[str] = field(
        default_factory=lambda: ["rag", "tools", "profile", "summary"]
    )


# ---------------------------------------------------------------------------
# Assembly result
# ---------------------------------------------------------------------------


@dataclass
class AssemblyResult:
    """Assembled context with per-source token accounting.

    Attributes:
        context:          Final assembled context string.
        token_breakdown:  Token count per source section.
        enforce_result:   :class:`~context_budget.EnforceResult` if enforcement ran.
        sources_included: Source names that made it into the context.
        sources_excluded: Source names that were dropped due to budget.
    """

    context: str
    token_breakdown: dict[str, int] = field(default_factory=dict)
    enforce_result: Optional[EnforceResult] = None
    sources_included: list[str] = field(default_factory=list)
    sources_excluded: list[str] = field(default_factory=list)

    @property
    def total_tokens(self) -> int:
        """Sum of tokens across all included sections."""
        return sum(self.token_breakdown.values())


# ---------------------------------------------------------------------------
# ContextAssembler
# ---------------------------------------------------------------------------


class ContextAssembler:
    """Assemble context for LLM calls from multiple sources.

    Template-driven, budget-aware, and optimizable.

    Example::

        budget    = ContextBudget(total_tokens=128_000)
        assembler = ContextAssembler(budget)

        result = assembler.assemble(
            template="You are a $role. Answer using the documents.",
            variables={"role": "support agent"},
            retrieved_docs=rag_docs,
            user_profile={"name": "Alice", "plan": "Pro"},
            query="How do I cancel my subscription?",
        )
        # result.context → final assembled string
        # result.token_breakdown → {"template": 45, "rag": 3200, "profile": 60}
        # result.sources_excluded → [] or ["summary", ...]
    """

    _SECTION_HEADERS: dict[str, str] = {
        "rag":     "## Retrieved Documents\n",
        "tools":   "## Tool Results\n",
        "profile": "## User Profile\n",
        "summary": "## Conversation Summary\n",
    }

    def __init__(
        self,
        budget: ContextBudget,
        optimizer: Optional[ContextOptimizer] = None,
        model: str = "gpt-4o",
    ) -> None:
        self.budget    = budget
        self.optimizer = optimizer or ContextOptimizer(model=model)
        self.model     = model

    # ------------------------------------------------------------------
    # Primary entry points
    # ------------------------------------------------------------------

    def assemble(
        self,
        template: str,
        variables: Optional[dict[str, str]] = None,
        retrieved_docs: Optional[list[dict]] = None,
        tool_results: Optional[list[dict]] = None,
        user_profile: Optional[dict] = None,
        conversation_summary: Optional[str] = None,
        query: str = "",
        optimize: bool = True,
    ) -> AssemblyResult:
        """Assemble the dynamic context for an LLM call.

        Sources are assembled in priority order:

        1. Retrieved documents (RAG)
        2. Tool execution results
        3. User profile / preferences
        4. Conversation summary
        5. Template variables

        Each source section is added while there is remaining budget.  When a
        section is too large to fit in full it is truncated to what fits.
        If there is no budget for a source it is excluded entirely.

        When *optimize* is ``True`` and a *query* is provided the RAG section
        is additionally structured and position-optimised before the full
        context is assembled.

        Args:
            template:             Python :class:`string.Template` template.
            variables:            Mapping for template substitution.
            retrieved_docs:       RAG result dicts with ``text``/``metadata``.
            tool_results:         Tool output dicts with ``tool``/``result``.
            user_profile:         Free-form user profile dict.
            conversation_summary: Plain-text summary of prior conversation turns.
            query:                User's question; used for optimisation.
            optimize:             Whether to apply :class:`ContextOptimizer`.

        Returns:
            :class:`AssemblyResult`.
        """
        variables = variables or {}

        # Render template
        try:
            rendered_template = Template(template).safe_substitute(variables)
        except Exception as exc:
            logger.warning("Template render failed: %s; using raw template", exc)
            rendered_template = template

        # Build source content map
        sections: dict[str, str] = {}

        if retrieved_docs:
            if optimize and query:
                structured = self.optimizer.structure_for_retrieval(retrieved_docs)
            else:
                structured = "\n\n".join(d.get("text", "") for d in retrieved_docs)
            sections["rag"] = structured

        if tool_results:
            tool_lines: list[str] = []
            for tr in tool_results:
                tool_name = tr.get("tool", "unknown")
                result = tr.get("result", "")
                if isinstance(result, (dict, list)):
                    result = json.dumps(result, indent=2)
                tool_lines.append(f"**{tool_name}**:\n{result}")
            sections["tools"] = "\n\n".join(tool_lines)

        if user_profile:
            lines = [f"- **{k}**: {v}" for k, v in user_profile.items()]
            sections["profile"] = "\n".join(lines)

        if conversation_summary:
            sections["summary"] = conversation_summary

        # Assemble in priority order within the dynamic_context budget
        dc_budget = self.budget.get_token_budget("dynamic_context")
        context_parts: list[str] = [rendered_template]
        token_breakdown: dict[str, int] = {
            "template": count_tokens(rendered_template, self.model)
        }
        sources_included: list[str] = []
        sources_excluded: list[str] = []

        for source in ["rag", "tools", "profile", "summary"]:
            if source not in sections:
                continue

            content = sections[source]
            header  = self._SECTION_HEADERS.get(source, f"## {source.title()}\n")
            section_text   = header + content
            section_tokens = count_tokens(section_text, self.model)
            used_so_far    = sum(token_breakdown.values())

            if used_so_far + section_tokens <= dc_budget:
                context_parts.append(section_text)
                token_breakdown[source] = section_tokens
                sources_included.append(source)
            else:
                # Try to fit a truncated version
                available = (
                    dc_budget
                    - used_so_far
                    - count_tokens(header, self.model)
                )
                if available > 100:
                    truncated    = self._truncate_to_tokens(content, available)
                    section_text = header + truncated
                    tok          = count_tokens(section_text, self.model)
                    context_parts.append(section_text)
                    token_breakdown[source] = tok
                    sources_included.append(source)
                    logger.warning(
                        "Source '%s' truncated: %d → %d tokens (budget %d)",
                        source, section_tokens, tok, dc_budget,
                    )
                else:
                    sources_excluded.append(source)
                    logger.warning(
                        "Source '%s' excluded: no budget remaining (%d / %d tokens used)",
                        source, used_so_far, dc_budget,
                    )

        final_context = "\n\n".join(context_parts)

        # Position-optimise the query-relevant RAG content
        if optimize and query and "rag" in sections:
            final_context = self.optimizer.optimize_for_query(final_context, query)

        return AssemblyResult(
            context=final_context,
            token_breakdown=token_breakdown,
            sources_included=sources_included,
            sources_excluded=sources_excluded,
        )

    def assemble_from_config(
        self,
        config: ContextConfig,
        retrieved_docs: Optional[list[dict]] = None,
        tool_results: Optional[list[dict]] = None,
        user_profile: Optional[dict] = None,
        conversation_summary: Optional[str] = None,
        query: str = "",
    ) -> AssemblyResult:
        """Assemble context from a declarative :class:`ContextConfig`.

        The config's ``include_sources`` gate and ``max_tokens_per_source``
        limits are applied before delegating to :meth:`assemble`.

        Args:
            config:               Assembly configuration.
            retrieved_docs:       RAG result dicts.
            tool_results:         Tool output dicts.
            user_profile:         Profile dict.
            conversation_summary: Summary string.
            query:                User's question.

        Returns:
            :class:`AssemblyResult`.
        """
        included = set(config.include_sources or ["rag", "tools", "profile", "summary"])

        filtered_docs    = retrieved_docs      if "rag"     in included else None
        filtered_tools   = tool_results        if "tools"   in included else None
        filtered_profile = user_profile        if "profile" in included else None
        filtered_summary = conversation_summary if "summary" in included else None

        # Apply per-source token limits
        for source, max_tok in (config.max_tokens_per_source or {}).items():
            if source == "rag" and filtered_docs is not None:
                filtered_docs = self._clip_docs_to_budget(filtered_docs, max_tok)

        return self.assemble(
            template=config.template,
            variables=config.template_vars or {},
            retrieved_docs=filtered_docs,
            tool_results=filtered_tools,
            user_profile=filtered_profile,
            conversation_summary=filtered_summary,
            query=query,
            optimize=(config.format == "markdown"),
        )

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _truncate_to_tokens(self, text: str, max_tokens: int) -> str:
        """Return *text* truncated to at most *max_tokens* tokens."""
        try:
            enc = tiktoken.encoding_for_model(self.model)
        except KeyError:
            enc = tiktoken.get_encoding("cl100k_base")
        tokens = enc.encode(text)
        if len(tokens) <= max_tokens:
            return text
        return enc.decode(tokens[:max_tokens])

    def _clip_docs_to_budget(
        self, docs: list[dict], max_tokens: int
    ) -> list[dict]:
        """Return the subset of *docs* that fits within *max_tokens*."""
        kept: list[dict] = []
        used = 0
        for doc in docs:
            t = count_tokens(doc.get("text", ""), self.model)
            if used + t <= max_tokens:
                kept.append(doc)
                used += t
            else:
                break
        return kept


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------


def _build_demo_rag_docs() -> list[dict]:
    return [
        {
            "text": (
                "## Account Management\n"
                "To update your account email, navigate to Settings → Account → Email "
                "Address.  A verification email will be sent to your new address."
            ),
            "metadata": {"source": "help-center/account-management.md", "score": 0.91},
        },
        {
            "text": (
                "## Billing & Subscriptions\n"
                "Invoices are generated on the 1st of each month.  Download them from "
                "Settings → Billing → Invoice History.  Invoices are retained for 7 years."
            ),
            "metadata": {"source": "help-center/billing.md", "score": 0.88},
        },
        {
            "text": (
                "## Cancellation Policy\n"
                "You may cancel your subscription at any time.  Cancellation takes effect "
                "at the end of the current billing period.  No refunds for partial months."
            ),
            "metadata": {"source": "help-center/cancellation.md", "score": 0.85},
        },
        {
            "text": (
                "## Data Export\n"
                "Export all your data as a ZIP archive from Settings → Privacy → Export "
                "Data.  The export includes messages, files, and account metadata."
            ),
            "metadata": {"source": "help-center/data-export.md", "score": 0.80},
        },
        {
            "text": (
                "## Two-Factor Authentication\n"
                "Enable 2FA in Settings → Security.  We support TOTP apps (Google "
                "Authenticator, Authy) and hardware security keys (FIDO2/WebAuthn)."
            ),
            "metadata": {"source": "help-center/security.md", "score": 0.78},
        },
    ]


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    print("=== Context Assembler Demo ===\n")

    budget    = ContextBudget(total_tokens=16_000)
    assembler = ContextAssembler(budget)

    rag_docs     = _build_demo_rag_docs()
    user_profile = {
        "name":         "Alice Johnson",
        "plan":         "Pro (annual)",
        "member_since": "2022-03-15",
        "open_tickets": "2",
    }
    prior_summary = (
        "Alice asked how to change her password (resolved: Settings → Security) "
        "and enquired about invoice downloads (resolved: Settings → Billing)."
    )
    tool_results = [
        {
            "tool": "lookup_account",
            "result": {
                "user_id":      "usr_alice_123",
                "subscription": "pro_annual",
                "next_invoice": "2026-06-01",
                "2fa_enabled":  True,
            },
        }
    ]

    query = "How do I cancel my subscription and download my last invoice?"

    result = assembler.assemble(
        template=(
            "You are a $role support agent. "
            "Answer the user's question using only the provided documents.\n"
        ),
        variables={"role": "customer"},
        retrieved_docs=rag_docs,
        tool_results=tool_results,
        user_profile=user_profile,
        conversation_summary=prior_summary,
        query=query,
    )

    print(f"Sources included : {result.sources_included}")
    print(f"Sources excluded : {result.sources_excluded}")
    print(f"Total tokens     : {result.total_tokens:,d}")
    print("\nToken breakdown by source:")
    for src, tok in result.token_breakdown.items():
        print(f"  {src:10s}: {tok:5,d} tokens")

    print("\n--- Assembled context (first 1 000 chars) ---")
    print(result.context[:1_000])

    print("\n\n--- assemble_from_config() demo ---")
    config = ContextConfig(
        template="You are a $role. Answer based on the documents only.",
        template_vars={"role": "support agent"},
        include_sources=["rag", "profile"],
        max_tokens_per_source={"rag": 2_000},
        format="markdown",
    )
    cfg_result = assembler.assemble_from_config(
        config,
        retrieved_docs=rag_docs,
        user_profile=user_profile,
        query=query,
    )
    print(f"Config result total tokens : {cfg_result.total_tokens:,d}")
    print(f"Sources included           : {cfg_result.sources_included}")
